### PR TITLE
VMware: Add writing new tests subsection to vmware_guidelines

### PR DESCRIPTION
**Pull request by jillr**
Specifically address use of the prepare_vmware_tests role
Point to common vars for use by test writers

##### SUMMARY
Add info about common vars and prepare role to aid test writers

##### ISSUE TYPE
- Docs Pull Request

##### COMPONENT NAME
docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst

### CODE REVIEW

**Comment by jillr** [comment]
cc @pgbidkar

File: docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst [hunk 1]
```diff
@@ -181,6 +181,24 @@ Depending upon the functionality provided by ESXi or vCenter, some modules can s
...
+The resources defined there are automatically created by importing that role are the start of your test:
+
+.. code-block:: yaml
+
+  - import_role:
+      name: prepare_vmware_tests
+    vars:
+      setup_datacenter: true
+
+This will give you a ready to use cluster, datacenter, datastores, folder, switch, dvswitch, ESXi hosts, and VMs.
```
Comment (pgbidkar) [inline docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst:200]: Can we mention the name of the each parameter in doc? ex: datacenter: {{dc1}}, cluster: {{ccr1}}.

File: docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst [hunk 2]
```diff
@@ -181,6 +181,24 @@ Depending upon the functionality provided by ESXi or vCenter, some modules can s
...
 ----------------
 
+Writing new tests
+~~~~~~~~~~~~~~~~~
+
+If you are writing a new collection of integration tests, there are a few VMware-specific things to note beyond
+the standard Ansible :ref:`integration testing<testing_integration>` process.
+
+The test-suite uses a set of common, pre-defined vars located in the :file:`test/integration/targets/prepare_vmware_tests/` role.
+The resources defined there are automatically created by importing that role are the start of your test:
```
Comment (Akasurde) [inline docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst]:
```suggestion
The resources defined there are automatically created by importing that role at the start of your test:
```

File: docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst (hunk 1 above)
Comment (jillr) [reply docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst:200]: My inclination is to just link to the role vars in the repo, and not duplicate them here as there would be a risk of them getting out of sync if we update the role. What do you think @goneri?  I don't think the vars will change often, but at the same time the less often we touch them we might forget that we also need to update these docs.

**Comment by agowa338** [comment]
Could you please also add a section regarding how one would start govcsim without running an integration test e.g. for "fiddling with the api" to write a new integration test?

Also is there a possibility to link a vcenter web ui to the govcsim? This would be very handy to get some insights and write checks for wired input (test locally what bs one can enter through the ui instead of the python/powercli api) validations (e.g. a "/" within the Datacenter name even breaks most native vmware apis, but it is still a valid character)

**Comment by Akasurde** [comment]
> Could you please also add a section regarding how one would start govcsim without running an integration test e.g. for "fiddling with the api" to write a new integration test?
> 

https://github.com/ansible/community/wiki/VMware%3A-govcsim

> Also is there a possibility to link a vcenter web ui to the govcsim? This would be very handy to get some insights and write checks for wired input (test locally what bs one can enter through the ui instead of the python/powercli api) validations (e.g. a "/" within the Datacenter name even breaks most native vmware apis (powercli/pyvmomi), but it is still a valid character)

Could explain this more ?

**Comment by agowa338** [comment]
> > Also is there a possibility to link a vcenter web ui to the govcsim? This would be very handy to get some insights and write checks for wired input (test locally what bs one can enter through the ui instead of the python/powercli api) validations (e.g. a "/" within the Datacenter name even breaks most native vmware apis (powercli/pyvmomi), but it is still a valid character)
> 
> Could explain this more ?

A few weeks ago, a college installed a vcenter for some internal testing and named the datacenter "DEV/Testing", this crashed everything. We were unable to access the new dev environment via Powercli (PowerCLI uses a path like "vis:/server/Datacenter/folder000/folder001/vm", but without escaping slashes and other special characters, therefore the path gets parsed incorrectly) and pyvmomi had similar problems.
Therefore it could be necessary to find these parsing issues.

Or as a stories:
- As a operator using the vmware ansible modules I want it to have the same features (deploy vms, rename stuff, ...) than the web ui.
- As a developer I want an integration test that covers every possible user input so that my module does not crash if a user uses a different frontend to rename/create stuff.

File: docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst (hunk 1 above)
Comment (pgbidkar) [reply docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst:200]: I agree.. thanks for clarification.

Comment (goneri) [reply docs/docsite/rst/dev_guide/platforms/vmware_guidelines.rst:200]: I don't have a strong opinion on this. The variable names are a bit confusing IMO (e.g: `ccs1` should  be `cluster1`), this for historical reason. I would prefer to fix that before.

In the meantime, I suggest we just point the user to https://github.com/ansible/ansible/tree/602aded6ecfdb20197bc6cfbc7b08ca01f5a7814/test/integration/targets/prepare_vmware_tests/vars

update: The `setup_*` variables should also be documented.

**Comment by goneri** [comment]
> > > Also is there a possibility to link a vcenter web ui to the govcsim? This would be very handy to get some insights and write checks for wired input (test locally what bs one can enter through the ui instead of the python/powercli api) validations (e.g. a "/" within the Datacenter name even breaks most native vmware apis (powercli/pyvmomi), but it is still a valid character)
> > 
> > 
> > Could explain this more ?
> 
> A few weeks ago, a college installed a vcenter for some internal testing and named the datacenter "DEV/Testing", this "crashed" everything except the ui. We were unable to access the new dev environment via Powercli (PowerCLI uses a path like "vis:/server/Datacenter/folder000/folder001/vm", but without escaping slashes and other special characters, therefore the path gets parsed incorrectly) and pyvmomi had similar problems.
> Therefore it could be necessary to find these parsing issues.
> 
> Or as a stories:
> 
>     * As a operator using the vmware ansible modules I want it to have the same features (deploy vms, rename stuff, ...) as the web ui does.
> 
>     * As a developer I want an integration test that covers every possible user input so that my module does not crash if a user uses a different frontend to rename/create stuff.

This is a broader issue that this documentation update. @agowa338, if you feel this is something that should address by the vmware_ modules, I suggest you open a feature request issue on the tracker.

**Review by goneri — APPROVED** [review]

**Pull request merged**